openedx/core/djangoapps/programs/utils.py

```diff
@@ -460,57 +461,86 @@ def _attach_course_run_upgrade_url(self, run_mode):
     def _attach_course_run_may_certify(self, run_mode):
         run_mode['may_certify'] = self.course_overview.may_certify()
 
-    def _check_enrollment_for_user(self, course_run):
-        applicable_seat_types = self.data['applicable_seat_types']
+    def _filter_out_courses_with_entitlements(self, courses):
+        """
+        Removes courses for which the current user already holds an applicable entitlement.
 
-        (enrollment_mode, active) = CourseEnrollment.enrollment_mode_for_user(
-            self.user,
-            CourseKey.from_string(course_run['key'])
-        )
+        Arguments:
+            courses (list): Containing dicts representing courses in a program
 
-        is_paid_seat = False
-        if enrollment_mode is not None and active is not None and active is True:
-            # Check all the applicable seat types
-            # this will also check for no-id-professional as professional
-            is_paid_seat = any(seat_type in enrollment_mode for seat_type in applicable_seat_types)
+        Returns:
+            A subset of the given list of course dicts
+        """
+        course_uuids = set(course['uuid'] for course in courses)
+        # Filter the entitlements' modes with a case-insensitive match against applicable seat_types
+        query = Q()
+        for seat_type in self.data['applicable_seat_types']:
+            query = query | Q(mode__iexact=seat_type)
+        query = Q(course_uuid__in=course_uuids) & query
+        entitlements = self.user.courseentitlement_set.filter(query)
+        courses_with_entitlements = set(str(entitlement.course_uuid) for entitlement in entitlements)
+        return [course for course in courses if course['uuid'] not in courses_with_entitlements]
+
+    def _filter_out_courses_with_enrollments(self, courses):
+        """
+        Removes courses for which the current user already holds an active and applicable enrollment
+        for one of that course's runs.
 
-        return is_paid_seat
+        Arguments:
+            courses (list): Containing dicts representing courses in a program
+
+        Returns:
+            A subset of the given list of course dicts
+        """
+        enrollments = self.user.courseenrollment_set.filter(
+            is_active=True,
+            mode__in=self.data['applicable_seat_types']
+        )
+        course_runs_with_enrollments = set(unicode(enrollment.course_id) for enrollment in enrollments)
+        courses_without_enrollments = []
+        for course in courses:
+            if len([run for run in course['course_runs'] if unicode(run['key']) in course_runs_with_enrollments]) == 0:
```

Comment:
Do we need to consider whether or not the run is published? Can you have an enrollment for an unpublished run?

Reply:
I believe unpublished is the course_run equivalent of expired

Reply:
Nit: This may be simplified by using `all`.

```python
if all(lambda run: unicode(run['key']) not in course_runs_with_enrollments, course['course_runs'])
```

Reply:
Nit: This might be simplified with `all`

```python
if all(run not in course_runs_with_enrollments for run in course['course_runs']):
```